src/transactions.rs

```diff
@@ -0,0 +1,456 @@
+    store: &'a mut TransactionsStore<L>,
+}
+
+impl<'a, L: Ledger + Serialize + DeserializeOwned> TransactionEditor<'a, L> {
+    /// Create a transaction editor.
+    fn new(store: &'a mut TransactionsStore<L>, transaction: Transaction<L>) -> Self {
+        Self { transaction, store }
+    }
+
+    /// Add fee change record to the transaction once it is certain
+    pub fn with_fee_change(mut self, amount: RecordAmount) -> Self {
+        self.transaction.fee_change = Some(amount);
+        self
+    }
+
+    /// Add asset change record to the transaction once it is certain
+    pub fn with_asset_change(mut self, amount: RecordAmount) -> Self {
+        self.transaction.asset_change = Some(amount);
+        self
+    }
+
+    /// Add the transaction receipt when it is recieved
+    pub fn with_receipt(mut self, receipt: TransactionReceipt<L>) -> Self {
+        self.transaction.receipt = Some(receipt);
+        self
+    }
+
+    /// Add the transaction hash, should be called after this transaction is committed
+    pub fn with_hash(mut self, hash: TransactionHash<L>) -> Self {
+        self.transaction.hash = Some(hash);
+        self
+    }
+
+    /// Add the UIDs of memos we are waiting on to complete the transaction
+    pub fn add_pending_uids(mut self, uids: &[u64]) -> Self {
+        for uid in uids {
+            self.transaction.pending_uids.insert(*uid);
+        }
+        self
+    }
+    /// remove a UID of a memo we were waiting because it was received
+    pub fn remove_pending_uid(mut self, uid: u64) -> Self {
+        self.transaction.pending_uids.remove(&uid);
+        self
+    }
+
+    /// Save the transaction to the store.
+    ///
+    /// Returns the stored transaction.
+    pub fn save(&mut self) -> Result<Transaction<L>, KeystoreError<L>> {
+        self.store.store(&self.transaction.uid, &self.transaction)?;
+        Ok(self.transaction.clone())
+    }
+}
+
+impl<'a, L: Ledger + Serialize + DeserializeOwned> Deref for TransactionEditor<'a, L> {
+    type Target = Transaction<L>;
+
+    fn deref(&self) -> &Transaction<L> {
+        &self.transaction
+    }
+}
+
+impl<'a, L: Ledger + Serialize + DeserializeOwned> DerefMut for TransactionEditor<'a, L> {
+    fn deref_mut(&mut self) -> &mut Transaction<L> {
+        &mut self.transaction
+    }
+}
+
+/// Transactions stored in an transactions store.
+pub struct Transactions<L: Ledger + Serialize + DeserializeOwned> {
+    /// A key-value store for transactions.
+    store: TransactionsStore<L>,
+    txn_by_hash: HashMap<TransactionHash<L>, TransactionUID<L>>,
+    expiring_txns: BTreeMap<u64, HashSet<TransactionUID<L>>>,
+    /// Maps pending UIDs to the Transaction<L>they come from
+    uids_awaiting_memos: HashMap<u64, TransactionUID<L>>,
+}
+
+impl<L: Ledger + Serialize + DeserializeOwned> Transactions<L> {
+    #![allow(dead_code)]
+
+    /// Load a transactions store.
+    pub fn new(store: TransactionsStore<L>) -> Result<Self, KeystoreError<L>> {
```

Comment:
Actually, can we make this take a `&mut AtomicStoreLoader` and construct the `TransactionsStore` itself? This will be more consistent with what @shenkeyao is working on with the `AtomicKeystoreStorage`, where basically the keystore session creates the loader and then passes it to each resource, which is then responsible for registering whatever underlying stores it needs.